include/rmm/cuda_event.hpp

```diff
@@ -0,0 +1,257 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime_api.h>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+
+#include <memory>
+
+namespace rmm {
+
+/**
+ * @brief Owning wrapper for a CUDA event.
+ *
+ * Provides RAII lifetime semantics for a CUDA event.
+ */
+class cuda_event {
+ public:
+  /**
+   * @brief Move constructor (default)
+   *
+   * A moved-from cuda_event is invalid and it is Undefined Behavior to call methods that access
+   * the owned event.
+   */
+  cuda_event(cuda_event&&) = default;
+  /**
+   * @brief Move copy assignment operator (default)
+   *
+   * A moved-from cuda_event is invalid and it is Undefined Behavior to call methods that access
+   * the owned event.
+   */
+  cuda_event& operator=(cuda_event&&) = default;
+  ~cuda_event()                       = default;
+  cuda_event(cuda_event const&)       = delete;  // Copying disallowed: one event one owner
+  cuda_event& operator=(cuda_event&) = delete;
+
+  /**
+   * @brief Construct a new cuda event object
+   *
+   * @throw rmm::cuda_error if event creation fails
+   */
+  cuda_event()
+    : event_{[]() {
+               cudaEvent_t* s = new cudaEvent_t;
+               RMM_CUDA_TRY(cudaEventCreate(s));
+               return s;
+             }(),
+             [](cudaEvent_t* s) {
+               RMM_ASSERT_CUDA_SUCCESS(cudaEventDestroy(*s));
+               delete s;
+             }}
+  {
+  }
+
+  /**
+   * @brief Construct a new cuda event object
+   *
+   * @throw rmm::cuda_error if event creation fails
+   */
+  cuda_event(unsigned int flags)
+    : event_{[flags]() {
+               cudaEvent_t* s = new cudaEvent_t;
+               RMM_CUDA_TRY(cudaEventCreateWithFlags(s, flags));
+               return s;
+             }(),
+             [](cudaEvent_t* s) {
+               RMM_ASSERT_CUDA_SUCCESS(cudaEventDestroy(*s));
+               delete s;
+             }}
+  {
+  }
+
+  /**
+   * @brief Returns true if the owned event is non-null
+   *
+   * @return true If the owned event has not been explicitly moved and is therefore non-null.
+   * @return false If the owned event has been explicitly moved and is therefore null.
+   */
+  bool is_valid() const { return event_ != nullptr; }
+
+  /**
+   * @brief Get the value of the wrapped CUDA event.
+   *
+   * @return cudaEvent_t The wrapped CUDA event.
+   */
+  cudaEvent_t value() const
+  {
+    RMM_LOGGING_ASSERT(is_valid());
+    return *event_;
+  }
+
+  /**
+   * @brief Explicit conversion to cudaEvent_t.
+   */
+  explicit operator cudaEvent_t() const noexcept { return value(); }
+
+  /**
+   * @brief Record the owned CUDA event in the given CUDA stream.
+   *
+   * Calls `cudaEventRecord()`.
+   *
+   * @throw rmm::cuda_error if event recording fails
+   */
+  void record(cudaStream_t stream) const { RMM_CUDA_TRY(cudaEventRecord(value(), stream)); }
```

Comment:
We shouldn't need any overloads for `cudaStream_t` as that defeats the purpose of having `cuda_stream_view`.